### PR TITLE
check if a module has a check method first

### PR DESCRIPTION
**Pull request by busterb**
Currently, if you run 'check' on a module that does not have a check method, it will first complain that you have not set the `RHOSTS` option, whether it's an exploit module or a scanner. Then, once you set RHOSTS (or whatever else it needs), it will then say 'Psych! I didn't have a method in the first place!'.

This switches that logic around so that it first alerts you that the module doesn't have support in the first place. It also similarizes more logic between aux and exploit for some future convergence possibility.

## Verification

- [x] Start `msfconsole`
- [x] `use auxiliary/scanner/http/http_put`
- [x] **Verify** the output looks something like this:
```
msf5 > use auxiliary/scanner/http/http_put
msf5 auxiliary(scanner/http/http_put) > run
[-] Auxiliary failed: Msf::OptionValidateError One or more options failed to validate: RHOSTS.
msf5 auxiliary(scanner/http/http_put) > check
[*] :80 - This module does not support check.
```

### CODE REVIEW

**Review by wvu — APPROVED** [review]
I like this. Looks fine, haven't tested.

File: lib/msf/base/simple/auxiliary.rb [hunk 1]
```diff
@@ -116,11 +116,12 @@ def self.check_simple(mod, opts, job_listener: Msf::Simple::NoopJobListener.inst
       mod.init_ui(opts['LocalInput'], opts['LocalOutput'])
     end
 
+    return Msf::Exploit::CheckCode::Unsupported unless mod.has_check?
```
Comment (adfoster-r7) [inline lib/msf/base/simple/auxiliary.rb:119]: Sanity question, does this impact the RPC service by returning immediately without a result? 🤔

Comment (adfoster-r7) [reply lib/msf/base/simple/auxiliary.rb:119]: Just asking as I believe this would impact the scenario of the RPC service expecting jobs to be asynchronous, and this method would normally return an array `[run_uuid, mod.job_id]` or raise a validation error in those scenarios. I think we might be impacting that contract.

Comment (wvu-r7) [reply lib/msf/base/simple/auxiliary.rb:119]: Did you mean `_check_auxiliary`? But yeah, I think the RPC interface expects an array. Good catch.

Comment (wvu-r7) [reply lib/msf/base/simple/auxiliary.rb:119]: Hold on, I'm verifying.

Comment (wvu-r7) [reply lib/msf/base/simple/auxiliary.rb:119]: With this patch:

```
>> rpc.call('module.check', 'auxiliary', 'scanner/http/http_put', {})
Traceback (most recent call last):
        8: from ./msfrpc:92:in `<main>'
        7: from /rapid7/metasploit-framework/lib/rex/ui/text/irb_shell.rb:52:in `run'
        6: from /rapid7/metasploit-framework/lib/rex/ui/text/irb_shell.rb:52:in `catch'
        5: from /rapid7/metasploit-framework/lib/rex/ui/text/irb_shell.rb:53:in `block in run'
        4: from (irb):1:in `<main>'
        3: from (irb):2:in `rescue in <main>'
        2: from /rapid7/metasploit-framework/lib/msf/core/rpc/v10/client.rb:105:in `call'
        1: from /rapid7/metasploit-framework/lib/msf/core/rpc/v10/client.rb:163:in `send_rpc_request'
RuntimeError (nil)
>>
```

Comment (wvu-r7) [reply lib/msf/base/simple/auxiliary.rb:119]: `master`:

```
>> rpc.call('module.check', 'auxiliary', 'auxiliary/scanner/http/http_put', {})
Traceback (most recent call last):
        7: from ./msfrpc:92:in `<main>'
        6: from /rapid7/metasploit-framework/lib/rex/ui/text/irb_shell.rb:52:in `run'
        5: from /rapid7/metasploit-framework/lib/rex/ui/text/irb_shell.rb:52:in `catch'
        4: from /rapid7/metasploit-framework/lib/rex/ui/text/irb_shell.rb:53:in `block in run'
        3: from (irb):1:in `<main>'
        2: from /rapid7/metasploit-framework/lib/msf/core/rpc/v10/client.rb:105:in `call'
        1: from /rapid7/metasploit-framework/lib/msf/core/rpc/v10/client.rb:158:in `send_rpc_request'
Msf::RPC::ServerException (Msf::OptionValidateError One or more options failed to validate: RHOSTS. ["lib/msf/core/module/options.rb:21:in `validate'", "lib/msf/base/simple/auxiliary.rb:121:in `check_simple'", "lib/msf/core/rpc/v10/rpc_module.rb:769:in `_check_auxiliary'", "lib/msf/core/rpc/v10/rpc_module.rb:518:in `rpc_check'", "lib/msf/core/rpc/v10/service.rb:155:in `block in process'", "lib/ruby/2.6.0/timeout.rb:93:in `block in timeout'", "lib/ruby/2.6.0/timeout.rb:33:in `block in catch'", "lib/ruby/2.6.0/timeout.rb:33:in `catch'", "lib/ruby/2.6.0/timeout.rb:33:in `catch'", "lib/ruby/2.6.0/timeout.rb:108:in `timeout'", "lib/msf/core/rpc/v10/service.rb:155:in `process'", "lib/msf/core/rpc/v10/service.rb:93:in `on_request_uri'", "lib/msf/core/rpc/v10/service.rb:74:in `block in start'", "lib/rex/proto/http/handler/proc.rb:38:in `on_request'", "lib/rex/proto/http/server.rb:368:in `dispatch_request'", "lib/rex/proto/http/server.rb:302:in `on_client_data'", "lib/rex/proto/http/server.rb:161:in `block in start'", "lib/rex/io/stream_server.rb:48:in `on_client_data'", "lib/rex/io/stream_server.rb:199:in `block in monitor_clients'", "lib/rex/io/stream_server.rb:197:in `each'", "lib/rex/io/stream_server.rb:197:in `monitor_clients'", "lib/rex/io/stream_server.rb:73:in `block in start'", "lib/rex/thread_factory.rb:22:in `block in spawn'", "lib/msf/core/thread_manager.rb:106:in `block in spawn'"])
>>
```

```
>> rpc.call('module.check', 'auxiliary', 'auxiliary/scanner/http/http_put', 'RHOSTS' => '127.0.0.1')
=> {"job_id"=>0, "uuid"=>"WOmSQIJyUG35uhnXoMradCTC"}
>>
```

Comment (wvu-r7) [reply lib/msf/base/simple/auxiliary.rb:119]: This patch:

```
>> rpc.call('module.check', 'auxiliary', 'auxiliary/scanner/http/http_put', {})
Traceback (most recent call last):
        7: from ./msfrpc:92:in `<main>'
        6: from /rapid7/metasploit-framework/lib/rex/ui/text/irb_shell.rb:52:in `run'
        5: from /rapid7/metasploit-framework/lib/rex/ui/text/irb_shell.rb:52:in `catch'
        4: from /rapid7/metasploit-framework/lib/rex/ui/text/irb_shell.rb:53:in `block in run'
        3: from (irb):1:in `<main>'
        2: from /rapid7/metasploit-framework/lib/msf/core/rpc/v10/client.rb:105:in `call'
        1: from /rapid7/metasploit-framework/lib/msf/core/rpc/v10/client.rb:163:in `send_rpc_request'
RuntimeError (nil)
>>
```

```
>> rpc.call('module.check', 'auxiliary', 'auxiliary/scanner/http/http_put', 'RHOSTS' => '127.0.0.1')
Traceback (most recent call last):
        8: from ./msfrpc:92:in `<main>'
        7: from /rapid7/metasploit-framework/lib/rex/ui/text/irb_shell.rb:52:in `run'
        6: from /rapid7/metasploit-framework/lib/rex/ui/text/irb_shell.rb:52:in `catch'
        5: from /rapid7/metasploit-framework/lib/rex/ui/text/irb_shell.rb:53:in `block in run'
        4: from (irb):1:in `<main>'
        3: from (irb):2:in `rescue in <main>'
        2: from /rapid7/metasploit-framework/lib/msf/core/rpc/v10/client.rb:105:in `call'
        1: from /rapid7/metasploit-framework/lib/msf/core/rpc/v10/client.rb:163:in `send_rpc_request'
RuntimeError (nil)
>>
```

Yeah, broken.

Comment (wvu-r7) [reply lib/msf/base/simple/auxiliary.rb:119]: Thanks for saving the day, @adfoster-r7!

Comment (adfoster-r7) [reply lib/msf/base/simple/auxiliary.rb:119]: I mean, technically it didn't _break_ as I expected it to do:

Request:
```
{
    "jsonrpc": "2.0",
    "id": "55919681-fb98-4efd-8354-f7e30fb556db",
    "method": "module.check",
    "params": [
        "auxiliary",
        "scanner/http/http_put",
        {
            "RHOSTS": "127.0.0.1",
            "RPORT": "3389"
        }
    ]
}
```

Response:

```
{
  "jsonrpc": "2.0",
  "result": {
    "job_id": null,
    "uuid": {
      "code": "unsupported",
      "message": "This module does not support check.",
      "reason": null
    }
  },
  "id": "55919681-fb98-4efd-8354-f7e30fb556db"
}
```

Looks like the UUID comes through as the module check guard clause that's been added

**Review by wvu — CHANGES_REQUESTED** [review]
Let's fix the RPC support first.

File: lib/msf/base/simple/auxiliary.rb (hunk 1 above)
Comment (wvu-r7) [reply lib/msf/base/simple/auxiliary.rb:119]: Prototyping how it probably should look:

```
>> rpc.call('module.check', 'auxiliary', 'auxiliary/scanner/http/http_put', {})
Traceback (most recent call last):
        8: from ./msfrpc:92:in `<main>'
        7: from /rapid7/metasploit-framework/lib/rex/ui/text/irb_shell.rb:52:in `run'
        6: from /rapid7/metasploit-framework/lib/rex/ui/text/irb_shell.rb:52:in `catch'
        5: from /rapid7/metasploit-framework/lib/rex/ui/text/irb_shell.rb:53:in `block in run'
        4: from (irb):1:in `<main>'
        3: from (irb):2:in `rescue in <main>'
        2: from /rapid7/metasploit-framework/lib/msf/core/rpc/v10/client.rb:105:in `call'
        1: from /rapid7/metasploit-framework/lib/msf/core/rpc/v10/client.rb:158:in `send_rpc_request'
Msf::RPC::ServerException (Msf::ValidationError This module does not support check. ["lib/msf/base/simple/auxiliary.rb:119:in `check_simple'", "lib/msf/core/rpc/v10/rpc_module.rb:769:in `_check_auxiliary'", "lib/msf/core/rpc/v10/rpc_module.rb:518:in `rpc_check'", "lib/msf/core/rpc/v10/service.rb:155:in `block in process'", "lib/ruby/2.6.0/timeout.rb:93:in `block in timeout'", "lib/ruby/2.6.0/timeout.rb:33:in `block in catch'", "lib/ruby/2.6.0/timeout.rb:33:in `catch'", "lib/ruby/2.6.0/timeout.rb:33:in `catch'", "lib/ruby/2.6.0/timeout.rb:108:in `timeout'", "lib/msf/core/rpc/v10/service.rb:155:in `process'", "lib/msf/core/rpc/v10/service.rb:93:in `on_request_uri'", "lib/msf/core/rpc/v10/service.rb:74:in `block in start'", "lib/rex/proto/http/handler/proc.rb:38:in `on_request'", "lib/rex/proto/http/server.rb:368:in `dispatch_request'", "lib/rex/proto/http/server.rb:302:in `on_client_data'", "lib/rex/proto/http/server.rb:161:in `block in start'", "lib/rex/io/stream_server.rb:48:in `on_client_data'", "lib/rex/io/stream_server.rb:199:in `block in monitor_clients'", "lib/rex/io/stream_server.rb:197:in `each'", "lib/rex/io/stream_server.rb:197:in `monitor_clients'", "lib/rex/io/stream_server.rb:73:in `block in start'", "lib/rex/thread_factory.rb:22:in `block in spawn'", "lib/msf/core/thread_manager.rb:106:in `block in spawn'"])
>>
```

```
msf5 auxiliary(scanner/http/http_put) > check
[-] Check failed: Msf::ValidationError This module does not support check.
msf5 auxiliary(scanner/http/http_put) >
```

Comment (wvu-r7) [reply lib/msf/base/simple/auxiliary.rb:119]: Thoughts?

Comment (wvu-r7) [reply lib/msf/base/simple/auxiliary.rb:119]: Tbh, we have to rely on a `raise` instead of  a `return` here as normally expected.

Comment (wvu-r7) [reply lib/msf/base/simple/auxiliary.rb:119]: I feel like it's either fix this or close this due to the side effects.

Comment (wvu-r7) [reply lib/msf/base/simple/auxiliary.rb:119]: https://github.com/busterb/metasploit-framework/pull/22

Comment (wvu-r7) [reply lib/msf/base/simple/auxiliary.rb]: It's actually a little weird that `CheckCode::Unsupported` is a thing you can return. I think it really should be raised. But I think it's fine to leave it because you can specify a reason now:

https://github.com/rapid7/metasploit-framework/blob/0417e88ff24bf05b8874c953bd91600f10186ba4/modules/exploits/linux/http/f5_bigip_tmui_rce.rb#L85-L87

Not that anyone else has used it for this.

Comment (busterb) [reply lib/msf/base/simple/auxiliary.rb]: Merged, thanks @wvu-r7

Comment (wvu-r7) [reply lib/msf/base/simple/auxiliary.rb]: If we want to make this a first-class exception, let me know. Otherwise, I am stealing bits and pieces.

**Review by wvu — APPROVED** [review]

**Comment by adfoster-r7** [comment]
@wvu This doesn't quite work just yet 👀 

For the JSON service, instead of getting the expected error response of `This module does not support check.`, the client instead gets:

image

It turns out the problem is with the error handling:

https://github.com/rapid7/metasploit-framework/blob/53a97f811832f22104d4dc8770f099f73c9ba4dd/lib/msf/core/rpc/json/dispatcher.rb#L114-L118

The code falls into the `rescue ArgumentError` clause when the new `Msf::ValidationError` exception is raised

```
From: /Users/adfoster/Documents/code/metasploit-framework/lib/msf/core/rpc/json/dispatcher.rb:115 Msf::RPC::JSON::Dispatcher#process_request:

     92: def process_request(request)
     93:   begin
     94:     if !validate_rpc_request(request)
     95:       response = self.class.create_error_response(InvalidRequest.new)
     96:       return response
     97:     end
     98: 
     99:     # dispatch method execution to command
    100:     result = @command.execute(request[:method], request[:params])
    101: 
    102:     # A Notification is a Request object without an "id" member. A Request
    103:     # object that is a Notification signifies the Client's lack of interest
    104:     # in the corresponding Response object, and as such no Response object
    105:     # needs to be returned to the client. The Server MUST NOT reply to a
    106:     # Notification, including those that are within a batch request.
    107:     if request.key?(:id)
    108:       response = self.class.create_success_response(result, request)
    109:     else
    110:       response = nil
    111:     end
    112: 
    113:     response
    114:   rescue ArgumentError
 => 115:     raise InvalidParams.new
    116:   rescue Msf::RPC::Exception => e
    117:     raise ApplicationServerError.new(e.message, data: { code: e.code })
    118:   end
    119: end
```

I thought something else was going wrong, but nope - the caught exception is indeed `Msf::ValidationError`:

```ruby
[8] pry(#<Msf::RPC::JSON::Dispatcher>)> $!
=> #<Msf::ValidationError: This module does not support check.>
[9] pry(#<Msf::RPC::JSON::Dispatcher>)> $!.class
=> Msf::ValidationError
```

The `Msf::ValidationError` actually extends `ArgumentError`:

```ruby
[10] pry(#<Msf::RPC::JSON::Dispatcher>)> $!.class.ancestors
=> [Msf::ValidationError,
 Msf::Exception,
 Rex::Exception,
 ArgumentError,
 StandardError,
 Exception,
 ActiveSupport::Dependencies::Blamable,
 Object,
 SNMP::BER,
 WindowsError::Win32,
 BSON::Object,
 Rex::Arch,
 ERB::Util,
 PP::ObjectMixin,
 PCAPRUB,
 ActiveSupport::Dependencies::Loadable,
 JSON::Ext::Generator::GeneratorMethods::Object,
 Kernel,
 BasicObject]
```

This means the user will get a cryptic error message back when a module doesn't have a check method:

```json
{
  "jsonrpc": "2.0",
  "error": {
    "code": -32602,
    "message": "Invalid method parameter(s)."
  },
  "id": "55919681-fb98-4efd-8354-f7e30fb556db"
}
```

This actually happens for incorrectly configured modules too, i.e. if `mod.validate` throws an `Msf::OptionValidation` exception for RHOSTS not being valid you get the same cryptic error:

image

I don't think this is intentional. I can create a separate ticket to make that error handling better, and we can ship this independently if overall it's making the world a better place.

---

TL:DR The existing json rpc service's error handling doesn't deal with the new `raise Msf::ValidationError` functionality correctly, but that seems like existing behaviour.

**Comment by wvu-r7** [comment]
Booooooo! Thanks for verifying.

File: lib/msf/base/simple/auxiliary.rb (hunk 1 above)
Comment (wvu-r7) [reply lib/msf/base/simple/auxiliary.rb]: @adfoster-r7

**Review by wvu — CHANGES_REQUESTED** [review]
Fix JSON-RPC.

**Comment by adfoster-r7** [comment]
I didn't want to scope creep this pull request too much, but I've updated the RPC service to be a little less rough around the edges. Hopefully.

https://github.com/busterb/metasploit-framework/pull/23

**Review by wvu — APPROVED** [review]

**Comment by adfoster-r7** [comment]
## Release Notes

Updates the check command to notify the user sooner if the module does not implement a check method. Specifically, users no longer need to provide valid module options before being finding out that a module does not implement a check method.
